tests/tests/htlc_tests.cpp

```diff
@@ -177,6 +177,448 @@ try {
+
+   int64_t bobs_balance = get_balance(bob, asset_id_type()(db));
+   // make sure Bob (or anyone) can see the details of the transaction
+   graphene::app::database_api db_api(db);
+   auto obj = db_api.get_objects( {alice_htlc_id }).front();
+   graphene::chain::htlc_object htlc = obj.template as<graphene::chain::htlc_object>(GRAPHENE_MAX_NESTED_OBJECTS);
+   BOOST_CHECK( htlc.memo.valid() );
+   
+   // grab number of history objects to make sure everyone gets notified
+   size_t alice_num_history = get_operation_history(alice_id).size();
+   size_t bob_num_history = get_operation_history(bob_id).size();
+   size_t joker_num_history = get_operation_history(joker_id).size();
+
+   // joker sends a redeem operation with a bad hash
+   {
+      std::vector<char> bad_pre_image{ 0x00, 0x01, 0x02 };
+      graphene::chain::htlc_redeem_operation update_operation;
+      update_operation.redeemer = joker_id;
+      update_operation.htlc_id = alice_htlc_id;
+      update_operation.preimage = bad_pre_image;
+      update_operation.fee = db.current_fee_schedule().calculate_fee( update_operation );
+      trx.operations.push_back( update_operation );
+      sign(trx, joker_private_key);
+      GRAPHENE_CHECK_THROW( PUSH_TX( db, trx, ~0 ), fc::exception );
+      trx.clear();      
+   }
+
+   // joker sends a redeem operation to claim the funds for bob
+   {
+      graphene::chain::htlc_redeem_operation update_operation;
+      update_operation.redeemer = joker_id;
+      update_operation.htlc_id = alice_htlc_id;
+      update_operation.preimage = pre_image;
+      update_operation.fee = db.current_fee_schedule().calculate_fee( update_operation );
+      trx.operations.push_back( update_operation );
+      sign(trx, joker_private_key);
+      PUSH_TX( db, trx, ~0 );
+      generate_block();
+      trx.clear();
+   }
+   // verify funds end up in Bob's account (3)
+   BOOST_CHECK_EQUAL( get_balance(bob_id,   graphene::chain::asset_id_type()), bobs_balance + 3 * GRAPHENE_BLOCKCHAIN_PRECISION );
+   // verify funds remain out of Alice's acount ( 100 - 3 - transaction fee )
+   BOOST_CHECK_EQUAL( get_balance(alice_id, graphene::chain::asset_id_type()), 9057813 );
+   // verify all three get notified
+   std::vector<operation_history_object> history = get_operation_history(alice_id);
+   BOOST_CHECK_EQUAL( history.size(), alice_num_history + 1);
+   display_operation_history{ history, alice };
+   history = get_operation_history(bob_id);
+   BOOST_CHECK_EQUAL( history.size(), bob_num_history + 1);
+   display_operation_history{ history, bob };
+   history = get_operation_history(joker_id);
+   BOOST_CHECK_EQUAL( history.size(), joker_num_history + 1);
+   display_operation_history{ history, joker };
+
+   // Alice attempts to put a proposal out with a preimage size of 0 (which is valid)
+   {
+      graphene::chain::htlc_create_operation create_operation;
+      BOOST_TEST_MESSAGE("Alice (who has 100 coins, is transferring 3 coins to Bob");
```

Comment:
Alice no longer has 100 coins. Also this test case is not about direct transfers.